### PR TITLE
[#IP-318] Update fp-ts lib from 1.x to 2.x

### PR DESCRIPTION
**Pull request by fabriziopapi**

### CODE REVIEW

File: src/tasks.ts [hunk 1]
```diff
@@ -1,18 +1,16 @@
...
  * Returns a Task that resolves to a value after a delay.
  */
 export const delayTask = <A>(n: Millisecond, a: A): Task<A> =>
-  new Task<A>(
-    () =>
-      new Promise<A>(resolve => {
-        setTimeout(() => resolve(a), n);
-      })
-  );
+  T.delay(n)(T.of(a));
```
Comment (AleDore) [inline src/tasks.ts:13]: nice ;)

**Review by AleDore — COMMENTED** [review]
Nice job! We have to check if we can safely remove `OptionEither` and `OptionValidation`, since they are not used anymore in our codebase :)

File: src/tasks.ts [hunk 2]
```diff
@@ -75,29 +73,38 @@ export const withRetries = <E, T>(
...
+            const delay = TE.of(
+              pipe(
+                delayTask(backoff(count), true),
+                T.map(d => right(d))
+              )
+            );
+            return pipe(
+              delay,
+              TE.chain(() => runTaskOnce(count + 1, currentTask))
+            );
```
Comment (balanza) [inline src/tasks.ts:95]: It seems to me that there's no need to lift a Task to a TaskEither, we can just write
```suggestion
            return pipe(
              delayTask(backoff(count), true),
              T.map(d => right(d)),
              TE.chain(() => runTaskOnce(count + 1, currentTask))
            );
```

File: src/index.ts [hunk 3]
```diff
@@ -56,11 +56,11 @@ export { pot };
...
 
-import * as OptionEither from "./OptionEither";
-export { OptionEither };
+// import * as OptionEither from "./OptionEither";
+// export { OptionEither };
 
-import * as OptionValidation from "./OptionValidation";
-export { OptionValidation };
+// import * as OptionValidation from "./OptionValidation";
+// export { OptionValidation };
```
Comment (balanza) [inline src/index.ts:63]: I think those modules aren't much used in our codebase, let's think on just remove them

**Comment by balanza** [comment]
> Nice job! We have to check if we can safely remove `OptionEither` and `OptionValidation`, since they are not used anymore in our codebase :)

I think we just need to understand if they have value or not (imho: they don't). Then we can safely remove them, and they're not much used in our codebase and we are in a breaking-change context already.

**Comment by balanza** [comment]
@fabriziopapi very good job, thanks! 
I haven't explored `2.x` syntax too much and I must say I like it a lot, I think we can really improve our codebase.

I just left a couple of minor comments, but the overall PR can be merged IMHO.

What's left?

File: src/tasks.ts (hunk 1 above)
Comment (gunzip) [reply src/tasks.ts:13]: https://github.com/gcanti/fp-ts/blob/master/src/Task.ts#L85

Comment (gunzip) [reply src/tasks.ts:13]: Since we now have a native delay method I'm wondering if we can just get rid of this wrapper.

Comment (balanza) [reply src/tasks.ts:13]: @gunzip do you mean [`timersPromises.setTimeout`](https://nodejs.org/docs/latest-v16.x/api/timers.html#timers_timerspromises_settimeout_delay_value_options)? They are available from Node 16

File: .node-version [hunk 4]
```diff
@@ -1 +1 @@
-10.14.1
+12.19.1
```
[the remark this answers is not in the document]
Comment (fabriziopapi) [reply .node-version:1]: Yes, the node update was required by some io-ts/fp-ts dependencies. I've fixed the the packege as suggested :-)

File: package.json [hunk 5]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@pagopa/ts-commons",
-  "version": "9.6.0",
+  "version": "10.0.0",
```
[the remark this answers is not in the document]
Comment (fabriziopapi) [reply package.json]: You're right, fixed!

**Comment by fabriziopapi** [comment]
> @fabriziopapi very good job, thanks!
> I haven't explored `2.x` syntax too much and I must say I like it a lot, I think we can really improve our codebase.
> 
> I just left a couple of minor comments, but the overall PR can be merged IMHO.
> 
> What's left?

I agree, the 2.x syntax has a lot of advantages.

I think we must choose if merge it in the master or release it from the brench as a new major "beta" version.

**Review by balanza — APPROVED** [review]

**Comment by balanza** [comment]
> I think we must choose if merge it in the master or release it from the branch as a new major "beta" version.

I think we should merge on master and then release new major, so that every next development will be on such major.

This PR is fine to me, when you're ready feel free to merge and release it.

**Pull request merged**